src/main/java/org/ethereum/vm/VM.java

```diff
@@ -0,0 +1,1053 @@
+                }
+
+                int memOffset = program.stackPop().intValueSafe();
+                int codeOffset = program.stackPop().intValueSafe();
+                int lengthData = program.stackPop().intValueSafe();
+
+                int sizeToBeCopied = (long) codeOffset + lengthData > fullCode.length
+                        ? (fullCode.length < codeOffset ? 0 : fullCode.length - codeOffset)
+                        : lengthData;
+
+                byte[] codeCopy = new byte[lengthData];
+
+                if (codeOffset < fullCode.length)
+                    System.arraycopy(fullCode, codeOffset, codeCopy, 0, sizeToBeCopied);
+
+                program.memorySave(memOffset, lengthData, codeCopy);
+                program.step();
+            }
+                break;
+            case GASPRICE: {
+                DataWord gasPrice = program.getGasPrice();
+
+                program.stackPush(gasPrice);
+                program.step();
+            }
+                break;
+
+            case BLOCKHASH: {
+
+                int blockIndex = program.stackPop().intValueSafe();
+
+                DataWord blockHash = program.getBlockHash(blockIndex);
+
+                program.stackPush(blockHash);
+                program.step();
+            }
+                break;
+            case COINBASE: {
+                DataWord coinbase = program.getCoinbase();
+
+                program.stackPush(coinbase);
+                program.step();
+            }
+                break;
+            case TIMESTAMP: {
+                DataWord timestamp = program.getTimestamp();
+
+                program.stackPush(timestamp);
+                program.step();
+            }
+                break;
+            case NUMBER: {
+                DataWord number = program.getNumber();
+
+                program.stackPush(number);
+                program.step();
+            }
+                break;
+            case DIFFICULTY: {
+                DataWord difficulty = program.getDifficulty();
+
+                program.stackPush(difficulty);
+                program.step();
+            }
+                break;
+            case GASLIMIT: {
+                DataWord gaslimit = program.getGasLimit();
+
+                program.stackPush(gaslimit);
+                program.step();
+            }
+                break;
+            case POP: {
+                program.stackPop();
+                program.step();
+            }
+                break;
+            case DUP1:
+            case DUP2:
+            case DUP3:
+            case DUP4:
+            case DUP5:
+            case DUP6:
+            case DUP7:
+            case DUP8:
+            case DUP9:
+            case DUP10:
+            case DUP11:
+            case DUP12:
+            case DUP13:
+            case DUP14:
+            case DUP15:
+            case DUP16: {
+                int n = op.val() - OpCode.DUP1.val() + 1;
```

Comment:
what is this `+ 1` magic number?

Reply:
I believe it's assuming the `DUP1` - `DUP16` are continuous and the `DUP1` is to push `1` byte.